sotodlib/core/flagman.py

```diff
@@ -0,0 +1,163 @@
+            flags: List of flags to buffer. Uses their names
+        """
+        if flags is None:
+            flags = self._fields
+        
+        for f in flags:
+            self[f].buffer(n_buffer)
+        
+    def buffered(self, n_buffer, flags=None):
+        """Return new FlagManager that has all the samps cuts buffered by n_buffer
+        Like with Ranges / Ranges Matrix, buffered returns new object
+        
+        Arguments:
+        
+            n_buffer: number of samples to buffer the samps cuts
+            flags: List of flags to buffer. Uses their names
+        """
+        new = self.copy()
+        new.buffer(n_buffer, flags)
+        return new
+        
+    def collapse(self, flags=None, method='union'):
+        """Collapse flags in the FlagManager together. If flags is None
+        it returns all flags
+        
+        Arguments:
+        
+            flags: List of flags to collapse together. Uses their names.
+            method: How to collapse the data. Accepts 'union','intersect',
+                        or function.
+        """
+        if flags is None:
+            flags = self._fields
+
+        to_reduce = [self._fields[f] for f in flags]
+        if len(flags)==0:
+            raise ValueError('Found zero flags to combine')
+            
+        out = RangesMatrix([Ranges(self.samps.count) for det in self.dets.vals])
+        
+        ## need to add out to prevent flag ordering from causing errors
+        ### (Ranges can't add to RangeMatrix, only other way around)
+        to_reduce[0] = out+to_reduce[0]
+        
+        if method == 'union': 
+            op = lambda x, y: x+y
+        elif method == 'intersect':
+            op = lambda x, y: x*y
+        else:
+            op = method
+        out = reduce(op, to_reduce)
+        return out
+    
+    def collapse_all(self, method='union'):
+        """same as collapse but for all fields, to be more verbose"""
+        return self.collapse(new_flag, list(self._fields.keys()), method)
+    
+    def reduce(self, new_flag, flags, method='union', keep=False):
+        """reduce multiple flags into a new flag by chosen method.
+        
+        Arguments:
+            
+            new_flag: Name of new flag
+            flags: list of flags to combine
+            method: see collapse
+            keep: if False, removes reduced flags from Manager
+        """
+        new = self.collapse(flags, method)
+        if keep and (new_flag in flags):
+            raise ValueError("keep == True would result in duplicate names.")
+        
+        # drop the fields if needed
+        if not keep: 
+            for f in flags: 
+                self.move(f, None)
+        
+        self.wrap(new_flag, new)
+        
+    def reduce_all(self, new_flag, method='union', keep=False):
+        """same as reduce but for all fields, to be more verbose"""
+        return self.reduce(new_flag, list(self._fields.keys()), method, keep)
```

Comment:
Once again... I think the verbosity would be better spent in the docstring of collapse() and reduce()!  For example, state what kind of object they return, and whether they modify self or not.